crates/dbsp/src/tutorial.rs

```diff
@@ -0,0 +1,744 @@
+//!     );
+//!     let mut reader = csv::Reader::from_path(path)?;
+//!     for result in reader.deserialize() {
+//!         let record: Record = result?;
+//!         println!("{:?}", record);
+//!     }
+//!     Ok(())
+//! }
+//! ```
+//!
+//! If we run this, then it prints the records in `Debug` format.  Here are the
+//! first few:
+//!
+//! ```text
+//! Record { location: "England", date: 2021-01-10, daily_vaccinations: None }
+//! Record { location: "England", date: 2021-01-11, daily_vaccinations: Some(140441) }
+//! Record { location: "England", date: 2021-01-12, daily_vaccinations: Some(164043) }
+//! Record { location: "England", date: 2021-01-13, daily_vaccinations: Some(192088) }
+//! Record { location: "England", date: 2021-01-14, daily_vaccinations: Some(213978) }
+//! ...
+//! ```
+//!
+//! ### Reading CSV into a DBSP circuit
+//!
+//! [`RootCircuit::build`] is the basic way to create a circuit in DBSP.
+//! A no-op use of it looks like this, which is also in `tutorial2.rs`:
+//!
+//! ```rust
+//! use anyhow::Result;
+//! use dbsp::RootCircuit;
+//!
+//! fn build_circuit(_circuit: &mut RootCircuit) -> Result<()> {
+//!     Ok(())
+//! }
+//!
+//! fn main() -> Result<()> {
+//!     let (_circuit, ()) = RootCircuit::build(build_circuit)?;
+//!     Ok(())
+//! }
+//! ```
+//!
+//! To start working with it, we can feed our records into a circuit through a
+//! "Z-set" ([`ZSet`]) input stream.  A "Z-set" is a set in
+//! which each item is associated with an integer weight.  In the context of
+//! changes to a data set, a positive weight represents an insertion and a
+//! negative weight represents a deletion.
+//!
+//! We create the Z-set input stream inside `build_circuit` using
+//! [`RootCircuit::add_input_zset`], which returns a [`Stream`] for further use
+//! in `build_circuit` and a [`CollectionHandle`] for the client to use to feed
+//! in data.  It looks like this:
+//!
+//! ```rust
+//! # use anyhow::Result;
+//! # use csv;
+//! # use dbsp::{CollectionHandle, OrdZSet, RootCircuit, ZSet};
+//! # use serde::Deserialize;
+//! # use size_of::SizeOf;
+//! # use time::Date;
+//! #
+//! # #[derive(Clone, Debug, Deserialize, Eq, PartialEq, Ord, PartialOrd, Hash, SizeOf)]
+//! # struct Record {
+//! #     location: String,
+//! #     date: Date,
+//! #     daily_vaccinations: Option<u64>,
+//! # }
+//! #
+//! fn build_circuit(circuit: &mut RootCircuit) -> Result<CollectionHandle<Record, isize>> {
+//!     let (input_stream, input_handle) = circuit.add_input_zset::<Record, isize>();
+//!     Ok(input_handle)
+//! }
+//!
+//! fn main() -> Result<()> {
+//!     let (circuit, mut input_handle) = RootCircuit::build(build_circuit)?;
+//!     // ...
+//! #     Ok(())
+//! }
+//! ```
+//!
+//! The best way to feed the records into `input_handle` is to collect them into
+//! a `Vec<(Record, isize)`, where `isize` is the Z-set weight.  All the weights
+//! can be 1, since we are inserting each of them.  We feed them in with
```

Comment:
I changed this to:

```rust
//! To start working with it, we can feed our records into a circuit through a
//! "Z-set" ([`ZSet`]) input stream.  A "Z-set" is a set in which each item is
//! associated with an integer weight.  In the context of changes to a data set,
//! positive weights represent insertions and negative weights represent
//! deletions.  The magnitude of the weight represents a count, so that a weight
//! of 1 represents an insertion of a single copy of a record, 2 represents two
//! copies, and so on, and similarly for negative weights and deletions.  Thus,
//! a Z-set represents changes to a multiset.
```